keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -49,120 +37,52 @@ class EnvironmentPromotionChecker(
-                true -> repository
-                  .getQueuedConstraintApprovals(deliveryConfig.name, environment.name)
-                  .toMutableSet()
-                false -> mutableSetOf()
-              }
-
-            val vetoedVersions: Set<String> =
-              (vetoedArtifacts[envPinKey(environment.name, artifact)]?.versions)
-                ?: emptySet()
+          return@forEach
+        }
+        deliveryConfig.environments.forEach { environment ->
+          // if pinned, we approve that version right away without doing anything else
+          // to fast track that deployment
+          if (pinnedEnvs.hasPinFor(environment.name, artifact)) {
+            // todo: https://github.com/spinnaker/keel/issues/1254
+            val pinnedVersion: String = pinnedEnvs.versionFor(environment.name, artifact) ?: return@forEach
+            // todo eb: this shouldn't be null, we just checked if a pin exists in the map above.
+            approveVersion(deliveryConfig, artifact, pinnedVersion, environment.name)
+            return@forEach
+          }
 
-            val version = when {
-              pinnedEnvs.hasPinFor(environment.name, artifact) -> {
-                hasPin = true
-                pinnedEnvs.versionFor(environment.name, artifact)
-              }
-              environment.constraints.isEmpty() && implicitConstraints.isEmpty() -> {
-                versions.firstOrNull { v ->
-                  !vetoedVersions.contains(v)
-                }
-              }
-              else -> {
-                versions
-                  .firstOrNull { v ->
-                    pendingVersionsToCheck.remove(v)
-                    queuedForApproval.remove(v)
-                    if (vetoedVersions.contains(v)) {
-                      false
-                    } else {
-                      /**
-                       * Only check stateful evaluators if all stateless evaluators pass. We don't
-                       * want to request judgement or deploy a canary for artifacts that aren't
-                       * deployed to a required environment or outside of an allowed time.
-                       */
-                      val passesConstraints =
-                        checkStatelessConstraints(artifact, deliveryConfig, v, environment) &&
-                          checkStatefulConstraints(artifact, deliveryConfig, v, environment)
-                      versionIsPending = when (environment.constraints.anyStateful) {
-                        true -> repository
-                          .constraintStateFor(deliveryConfig.name, environment.name, v)
-                          .any { it.status == PENDING }
-                        else -> false
-                      }
-                      passesConstraints || versionIsPending
-                    }
-                  }
+          // todo eb: should this be in another loop?
+          constraintRunner.checkEnvironment(
```

Comment:
Ah, I mean should the `EnvironmentConstraintRunner` actually just be called from the `CheckScheduler` like this method is, instead of being called by this method

Reply:
I see. That's a good question. I'm not sure either. 😕

Reply:
removing the to-do and also not moving it now, i think it's okay as is